Allow for checking field level annotations from DiffNode (getFieldAnn…

NagyGa1 opened the pull request:
…otations() functions added, similarly to getPropertyAnnotations()).

It is important specially for handling JPA classes where annotations can be at both places.

An alternative for the user of the library would be to use DiffNode%getParentNode()%getValueType() and DiffNode%getPropertyName() to do the same without the getFieldAnnotations() functions.

The client may even use something like Jandex (https://github.com/wildfly/jandex).

SQiShER commented on src/main/java/de/danielbechler/diff/node/DiffNode.java:
```diff
@@ -444,6 +444,29 @@ public final void visitParents(final Visitor visitor)
 	}
 
 	/**
+	 * If this node represents a bean property this method returns all annotations of its field.
```
To avoid confusion, it would be good to mention that this will only work when the field name matches the property name extracted from the getter.

NagyGa1 replied:
K, done

SQiShER commented on src/main/java/de/danielbechler/diff/introspection/PropertyAccessor.java:
```diff
@@ -84,6 +84,44 @@ public String getPropertyName()
+	 * @return
+	 */
+	private Set<Annotation> getFieldAnnotations(final Class<?> clazz) {
+		try {
```
The code format (line breaks) of the changes in this class is a little off.

NagyGa1 replied:
I can't get this one, what do you mean?

SQiShER commented:
Thanks for the pull request! The implementation looks good to me. :+1: 

Other than the small gripes from above, the only thing that's keeping me from merging it right away is the lack of tests. It would be nice to see some unit and/or integration tests, to avoid future changes from accidentally breaking this feature.

I'd like to see at least one integration test following the happy path via `ObjectDiffer.compare()`, to make sure the feature actually works. Just put a new Spock test class [here](https://github.com/SQiShER/java-object-diff/tree/master/src/integration-test/java/de/danielbechler/diff/node) and give it a feature related name followed by "IT".

I'd also like to see a couple unit tests for the public `getFieldAnnotation(Class<T>)` method, basically testing all the branches of the implementation:

- field exists in given path
- no such field, with superclass
- no such field, without superclass
- no such field, with superclass that has the field
- no such field, with superclass that does not have the field
- annotation found
- annotation not found
- etc.

You get the idea... :wink:

Other than that, great PR. I'm looking forward to merging it.

NagyGa1 commented:
Looking at the ITs, might take a while, need to learn Groovy first.

NagyGa1 commented:
...and Spock.

NagyGa1 commented:
Ok, adjusted javadocs, added unit tests on, let the line breaks (please clarify).

Not sure about integration test, the feature is not integrated at the moment, solely to be called directly on DiffNode. Have not looked at integration to comparisons or inclusions.

SQiShER commented:
> Ok, adjusted javadocs, added unit tests on, left the line breaks (please clarify).

Looks good to me. Regarding the line breaks: just put each curly brace on its own line.

> Should getPropertyAnnotations() marked deprecated, merged with getFieldAnnotations() under the name of getAnnotations(), then used in the integrated part of the library? Changes a few contracts slightly.

I wouldn't do that for now. I think it is more flexible the way it is. If it turns out to be a hassle, I'd prefer adding another convenience method that uses the other ones internally, but still keep the other ones public.

> Not sure about integration test, the feature is not integrated at the moment, solely to be called directly on DiffNode. Have not looked at integration to comparisons or inclusions.

Good point. Yet I always like to see a feature working by using it like an actual user, just to be sure. But I can add the test myself. I already wrote one yesterday to play around with the feature. So don't sweat it.

Just fix the line breaks, squash your commits into one and we're good to go. :+1:

NagyGa1 commented:
Thank you, will do it shortly.

I use the feature like this:
```java
package com.cpox.view.config.expimp;

import java.lang.annotation.Annotation;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.javatuples.Pair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import de.danielbechler.diff.inclusion.Inclusion;
import de.danielbechler.diff.inclusion.InclusionResolver;
import de.danielbechler.diff.node.DiffNode;

/**
 * Excludes given annotations.
 */
public class AnnotationExclusionResolver implements InclusionResolver {

	private static final Logger LOG = LoggerFactory
			.getLogger(AnnotationExclusionResolver.class);

	private final List<Class<?>> excludedAnnotations;

	private Map<Pair<Class<?>, String>, Inclusion> responseCache = new HashMap<>();

	public AnnotationExclusionResolver(final List<Class<?>> excludedAnnotations) {
		this.excludedAnnotations = excludedAnnotations;
	}

	@Override
	public Inclusion getInclusion(final DiffNode node) {

		// generate ID for cache retrieval
		final Pair<Class<?>, String> id = new Pair<Class<?>, String>(node
				.getParentNode().getValueType(), node.getPropertyName());

		// check from cache
		Inclusion response = responseCache.get(id);

		if (response == null) {
			// not in cache, calculate value
			response = calculateInclusion(node);
			responseCache.put(id, response);
			if (LOG.isDebugEnabled() && response == Inclusion.EXCLUDED) {
				LOG.debug("{} {}", id, response);
			}
		}

		return response;
	}

	private Inclusion calculateInclusion(final DiffNode node) {
		for (Class<?> excludedAnnotation : excludedAnnotations) {
			for (Annotation annotation : node.getPropertyAnnotations()) {
				if (annotation.annotationType().isAssignableFrom(
						excludedAnnotation)) {
					return Inclusion.EXCLUDED;
				}
			}
		}
		for (Class<?> excludedAnnotation : excludedAnnotations) {
			for (Annotation annotation : node.getFieldAnnotations()) {
				if (annotation.annotationType().isAssignableFrom(
						excludedAnnotation)) {
					return Inclusion.EXCLUDED;
				}
			}
		}
		return Inclusion.DEFAULT;
	}

	@Override
	public boolean enablesStrictIncludeMode() {
		return false;
	}
}
```

NagyGa1 commented:
Ok, done.

SQiShER commented:
Thanks for this PR and also for sharing your `AnnotationExclusionResolver`. It's great to see the `InclusionResolver` (which is a fairly new feature) being leveraged in such a good way!

SQiShER commented:
I'll release a new version over the weekend, so you won't have to use a snapshot for too long.

NagyGa1 commented:
Thank you very much.

NagyGa1 commented:
Sorry, I am not too experienced in GitHub: should I delete my branch / fork?

SQiShER commented:
@NagyGa1 Version 0.92.1 is now released and should soon (in about 2 hours) be available via Maven Central.

NagyGa1 commented:
Thank you very much.
You can't sleep?

On Sat, 6 Jun 2015 6:52 am Daniel Bechler <notifications@github.com> wrote:

> @NagyGa1 <https://github.com/NagyGa1> Version 0.92.1 is now released and
> should soon (in about 2 hours) be available via Maven Central.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SQiShER/java-object-diff/pull/134#issuecomment-109469771>
> .
>

SQiShER commented:
Nah, had too much caffeinated iced tea :smile:

NagyGa1 commented:
I am ashamed to admit drinking decaf coffee these days. :/

Anyway, tks for the fast release!

On Sat, 6 Jun 2015 6:56 am Daniel Bechler <notifications@github.com> wrote:

> Nah, had too much caffeinated iced tea [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SQiShER/java-object-diff/pull/134#issuecomment-109471449>
> .
>
